File: web/dart/blocks/chain.dart
/*
 * NetTango
 * Copyright (c) 2019 Michael S. Horn, Uri Wilensky, and Corey Brady
 *
 * Northwestern University
 * 2120 Campus Drive
 * Evanston, IL 60613
 * http://tidal.northwestern.edu
 * http://ccl.northwestern.edu

 * This project was funded in part by the National Science Foundation.
 * Any opinions, findings and conclusions or recommendations expressed in this
 * material are those of the author(s) and do not necessarily reflect the views
 * of the National Science Foundation (NSF).
 */
part of NetTango;

class Chain extends BlockCollection {

  static final FRAGMENT_HEIGHT = 20;

  int x = 0, y = 0;

  final CodeWorkspace workspace;

  int chainIndex;

  DivElement fragmentDiv;
  bool isDragOver = false;

  // TODO: Starter/Fragment should really be explicit values in the data model
  bool get isFragment => blocks.isEmpty || !blocks[0].required;

  Chain(CodeWorkspace this.workspace);

  DivElement draw(DragImage dragImage, int newChainIndex) {
    this.chainIndex = newChainIndex;

    fragmentDiv = new DivElement();
    fragmentDiv.classes.add("nt-fragment");
    final fragmentDropzone = Dropzone(fragmentDiv, acceptor: this.workspace.workspaceAcceptor);
    fragmentDropzone.onDrop.listen(drop);
    fragmentDropzone.onDragEnter.listen( (e) => isDragOver = true );
    fragmentDropzone.onDragLeave.listen( (e) => isDragOver = false );

    _div = new DivElement();
    _div.classes.add("nt-chain");

    if (blocks.isEmpty) {
      return _div;
    }

    for (int i = 0; i < blocks.length; i++) {
      Block block = blocks.elementAt(i);
      final dragData = BlockDragData.workspaceChain(chainIndex, i, blocks.skip(i + 1));
      block.draw(dragImage, dragData);
    }

    Chain.redrawChain(_div, blocks, false, fragmentDiv: fragmentDiv);

    updatePosition(this.x, this.y);

    return _div;
  }

  void updatePosition(int x, int y) {
    this.x = x;
    this.y = y;
    _div.style.left = "${x}px";
    _div.style.top  = "${y}px";
  }

  void resetDragData(int newChainIndex) {
    this.chainIndex = newChainIndex;
    for (int i = 0; i < blocks.length; i++) {
      Block block = blocks[i];
      block._dragData.resetWorkspaceChain(chainIndex, i, blocks.skip(i + 1));
      block.resetOwnedBlocksDragData();
    }
  }

  bool updateDragOver() {
    fragmentDiv.classes.remove("nt-drag-over");
    bool isHighlightHandled = false;
    for (Block block in blocks) {
      final blockResult = block.updateDragOver();
      isHighlightHandled = isHighlightHandled || blockResult;
    }
    if (isDragOver && !isHighlightHandled) {
      isHighlightHandled = true;
      fragmentDiv.classes.add("nt-drag-over");
    }
    return isHighlightHandled;
  }

  void clearDragOver() {
    fragmentDiv.classes.remove("nt-drag-over");
    isDragOver = false;
    for (Block block in blocks) {
      block.clearDragOver();
    }
  }

  static void redrawChain(DivElement div, List<Block> blocks, bool useClones, {DivElement fragmentDiv = null}) {
    div.innerHtml = "";
    if (blocks.first.required) {
      div.classes.add("nt-chain-starter");
      div.classes.remove("nt-chain-fragment");
    } else {
      div.classes.remove("nt-chain-starter");
      div.classes.add("nt-chain-fragment");
      if (fragmentDiv != null) {
        div.append(fragmentDiv);
      }
      final topNotch = Notch.draw(true, blocks.first);
      div.append(topNotch);
    }

    BlockCollection.appendBlocks(div, blocks, "nt-block", useClones: useClones);

    final bottomNotch = Notch.draw(false, blocks.last);
    div.append(bottomNotch);
  }

  void redrawBlocks() {
    for (int i = 0; i < blocks.length; i++) {
      Block block = blocks[i];
      block._dragData.resetWorkspaceChain(chainIndex, i, blocks.skip(i + 1));
      block.resetOwnedBlocksDragData();
    }
    Chain.redrawChain(_div, blocks, false, fragmentDiv: fragmentDiv);
  }

  Iterable<Block> removeBlocks(int blockIndex) {
    final removedBlocks = blocks.skip(blockIndex);
    blocks = blocks.take(blockIndex).toList();
    redrawBlocks();
    return removedBlocks;
  }

  void addBlocks(Iterable<Block> newBlocks) {
    insertBlocks(blocks.length, newBlocks);
  }

  void insertBlocks(int blockIndex, Iterable<Block> newBlocks) {
    blocks.insertAll(blockIndex, newBlocks);
    redrawBlocks();
  }

  void enableTopDropZone() {
    if (!isFragment) {
      return;
    }
    fragmentDiv.classes.add("show");
    final top = this.y.round() - FRAGMENT_HEIGHT;
    _div.style.top = "${top}px";
  }

  void disableTopDropZone() {
    fragmentDiv.classes.remove("show");
    final top = this.y.round();
    _div.style.top = "${top}px";
  }

  void drop(DropzoneEvent event) {
    DragAcceptor.wasHandled = true;

    final newBlocks = workspace.consumeDraggingBlocks();
    final newFirst  = newBlocks.elementAt(0);
    final offset = DragImage.getOffsetToRoot(this._div);
    final dropLocation = event.position - offset;
    this.y = this.y - FRAGMENT_HEIGHT + dropLocation.y.floor();
    insertBlocks(0, newBlocks);

    workspace.programChanged(new BlockChangedEvent(newFirst));
    workspace.disableTopDropZones();
  }

}

File: src/blocks/chain.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import interact from "interactjs"
import type { InteractEvent } from '@interactjs/core/InteractEvent'

import { BlockInstanceUI } from "./block-instance"
import { BlockCollection } from "./block-collection"
import { CodeWorkspaceUI } from "./code-workspace"
import { ChainAcceptor } from "./drag-drop/chain-acceptor"
import { ChainDragData } from "./drag-drop/drag-data/chain-drag-data"
import { DragListener } from "./drag-drop/drag-listener"
import { DragManager } from "./drag-drop/drag-manager"
import { BlockChangedEvent } from "./program-changed-event"
import { Chain } from "../types/types"
import { Cap } from "./baubles/cap"
import { ChainDraw } from "./chain-draw"

class ChainUI extends BlockCollection {

  static readonly FRAGMENT_HEIGHT = 40

  readonly c: Chain
  readonly workspace: CodeWorkspaceUI

  chainIndex: number

  fragmentDiv = document.createElement("div")

  get isFragment(): boolean { return this.blocks.length === 0 || !this.blocks[0].canBeStarter }

  constructor(c: Chain, workspace: CodeWorkspaceUI, chainIndex: number) {
    super(c.blocks, workspace)
    this.c = c
    this.workspace = workspace
    this.chainIndex = chainIndex
  }

  draw(newChainIndex: number): HTMLDivElement {
    this.chainIndex = newChainIndex

    this.fragmentDiv = document.createElement("div")
    this.fragmentDiv.classList.add("nt-fragment")
    const acceptor = new ChainAcceptor(this)
    const fragmentDropzone = interact(this.fragmentDiv).dropzone({
      accept: ".nt-menu-slot, .nt-block, .nt-cap, .nt-notch"
      , checker: (_1, _2, dropped) => acceptor.checker(dropped)
    })
    fragmentDropzone.on("drop", (e) => this.drop(e) )
    fragmentDropzone.on("dragenter", () => {
      this.fragmentDiv.classList.add("nt-drag-over")
    } )
    fragmentDropzone.on("dragleave", () => {
      this.fragmentDiv.classList.remove("nt-drag-over")
    } )

    this.div = document.createElement("div")
    this.div.classList.add("nt-chain")

    if (this.blocks.length === 0) {
      return this.div
    }

    for (var i = 0; i < this.blocks.length; i++) {
      const block = this.blocks[i]
      const dragData = new ChainDragData(block, this.chainIndex, i, this.blocks.slice(i + 1))
      block.draw(dragData)
    }

    ChainDraw.draw(Cap.draw, this.div, this.blocks, false, this.fragmentDiv)

    this.updatePosition(this.c.x, this.c.y)

    return this.div
  }

  updatePosition(x: number, y: number): void {
    this.c.x = x
    this.c.y = y
    this.div.style.left = `${x}px`
    this.div.style.top  = `${y}px`
  }

  resetDragData(newChainIndex: number): void {
    this.chainIndex = newChainIndex
    for (var i = 0; i < this.blocks.length; i++) {
      const block = this.blocks[i]
      block.dragData = new ChainDragData(block, this.chainIndex, i, this.blocks.slice(i + 1))
      block.resetOwnedBlocksDragData()
    }
  }

  redrawBlocks(): void {
    for (var i = 0; i < this.blocks.length; i++) {
      const block = this.blocks[i]
      block.dragData = new ChainDragData(block, this.chainIndex, i, this.blocks.slice(i + 1))
      block.resetOwnedBlocksDragData()
    }
    ChainDraw.draw(Cap.draw, this.div, this.blocks, false, this.fragmentDiv)
  }

  addBlocks(newBlocks: BlockInstanceUI[]): void {
    this.insertBlocks(this.blocks.length, newBlocks)
  }

  enableDropZones(): void {
    if (DragManager.isValidDrop(this.workspace.containerId, (dragState) => ChainAcceptor.isLandingSpot(this, dragState))) {
      this.div.classList.add("nt-allowed-drop")
    }

    if (this.isFragment) {
      this.fragmentDiv.classList.add("show")
      const top = Math.round(this.c.y) - ChainUI.FRAGMENT_HEIGHT
      this.div.style.top = `${top}px`
    }

    for (var block of this.blocks) {
      block.enableDropZones()
    }
  }

  disableDropZones(): void {
    this.div.classList.remove("nt-allowed-drop")

    this.fragmentDiv.classList.remove("show")
    const top = Math.round(this.c.y)
    this.div.style.top = `${top}px`

    for (var block of this.blocks) {
      block.disableDropZones()
    }
  }

  drop(event: InteractEvent): void {
    DragManager.drop( (newBlocks, dragStartOffset) => {
      this.fragmentDiv.classList.remove("nt-drag-over")
      const newFirst = newBlocks[0]
      const offset = DragListener.getOffsetToRoot(this.div)
      // The casts here are necessary I believe because the type defs are wrong, `dragEvent` does exist on the
      // `InteractEvent` when a drop occurs. -Jeremy B January 2020
      const dropY = ((event as any).dragEvent.page.y as number) - offset.y - dragStartOffset.y
      this.c.y = this.c.y - ChainUI.FRAGMENT_HEIGHT + Math.floor(dropY)
      this.insertBlocks(0, newBlocks)
      this.workspace.programChanged(new BlockChangedEvent(newFirst))
    })
  }

}

export { ChainUI as ChainUI }
